### PR TITLE
fleet combine amp dgc recompute meta optimizer

### PR DESCRIPTION
**Pull request by wangxicoding**
### PR types

Others
### PR changes

Others
### Describe

Refine amp&dgc&recompute&lars&lamb&local_sgd, make them can be an inner_opt of other meta optimizer.
Now the following strategies that can be combined：
``` python
import paddle.distributed.fleet as fleet
strategy = fleet.DistributedStrategy()

# 1. amp + recompute
strategy.amp = True
strategy.recompute = True

# 2. dgc + recompute
strategy.dgc = True
strategy.recompute = True

# 3. lars + recompute
strategy.lars = True
strategy.recompute = True

# 4. lamb + recompute
strategy.lamb = True
strategy.recompute = True

# 6. amp + localsgd
strategy.amp = True
strategy.localsgd = True

# 6. amp + adaptive_localsgd
strategy.amp = True
strategy.adaptive_localsgd = True

# 7. amp + lars
strategy.amp = True
strategy.lars = True

# 8. amp + lamb
strategy.amp = True
strategy.lamb = True

# amp + dgc still have a problem, not combine them for now.
# 9. need supply
```

### CODE REVIEW

**Review by ForFishes — DISMISSED** [review]
LGTM

File: python/paddle/distributed/fleet/meta_optimizers/recompute_optimizer.py [hunk 1]
```diff
@@ -18,24 +18,31 @@
...
         super(RecomputeOptimizer, self).__init__(optimizer)
-        #self.inner_opt = RO(optimizer)
         self.inner_opt = optimizer
-        self.wrapped_opt = RO(optimizer)
+        self.wrapped_opt = None
         # we do not allow meta optimizer to be inner optimizer currently
         self.meta_optimizers_white_list = [
             "LarsOptimizer",
             "LambOptimizer",
             "GradientMergeOptimizer",
```
Comment (mapingshuo) [inline python/paddle/distributed/fleet/meta_optimizers/recompute_optimizer.py]: help remove this from the `meta_optimizers_white_list` of RecomputeOptimizer.

**Review by mapingshuo — COMMENTED** [review]
LGTM for recomputeOptimizer.

File: python/paddle/distributed/fleet/meta_optimizers/recompute_optimizer.py (hunk 1 above)
Comment (wangxicoding) [reply python/paddle/distributed/fleet/meta_optimizers/recompute_optimizer.py]: removed

**Review by mapingshuo — APPROVED** [review]
LGTM

**Pull request merged**